lib/Runtime/Library/ForInObjectEnumerator.cpp

```diff
@@ -182,7 +207,32 @@ namespace Js
                 if (TestAndSetEnumerated(propertyId) //checks if the property is already enumerated or not
                     && (attributes & PropertyEnumerable))
                 {
-                    return currentIndex;
+                    bool propertyShadowed = false;
```

Comment:
That's right. For long prototype chain and most of the prototype objects are non built-in objects, this will make the lookup much slower. For this case, I think we could move firstPrototype up after a prototype has been enumerated instead of keeping it fixed for a given prototype chain. This will fix the perf that non-built objects inherits from built-in objects in prototype chain, but it will still be an issue if built-in objects are inherited from non built-in objects. Wondering whether the latter case is typical? If it is, we could add some heuristics like traverse the whole chain after some number of downward look-ups? I will do the former optimization. May do the latter if necessary. What do you think, @curtisman ?